File: src/05.py
import sys
import fileinput
from collections import defaultdict


def solve_1(pointPairs):
    grid = defaultdict(int)
    for pointPair in pointPairs:
        markLine(grid, pointPair)
    return len([k for k in grid if grid[k] > 1])


def solve_2(pointPairs):
    grid = defaultdict(int)
    for pointPair in pointPairs:
        markLine(grid, pointPair, True)
    return len([k for k in grid if grid[k] > 1])


def markLine(grid, pointPair, with_diagonals=False):
    (x1, y1), (x2, y2) = pointPair
    dx, dy = x2 - x1, y2 - y1
    if not with_diagonals and dx != 0 and dy != 0:
        return

    n = max(abs(dx), abs(dy))
    inc_x = (1 if dx > 0 else -1) if dx != 0 else 0
    inc_y = (1 if dy > 0 else -1) if dy != 0 else 0
    for i in range(0, n+1):
        p = (x1+i*inc_x, y1+i*inc_y)
        grid[p] += 1


if __name__ == '__main__' and not sys.flags.interactive:
    lines = [line.strip() for line in fileinput.input()]
    pairs = [line.split(' -> ') for line in lines]
    pointPairs = [[tuple([int(w) for w in s.split(',')]) for s in pair] for pair in pairs]
    print(solve_1(pointPairs))
    print(solve_2(pointPairs))

File: src/05.go
package main

import (
	"fmt"

	. "aoc2021/src/common"
)

func solve_05_1(lines []line) int {
	grid := map[point]int{}
	for _, line := range lines {
		if line.x1 == line.x2 || line.y1 == line.y2 {
			markLine(line, grid)
		}
	}
	return countResult(grid)
}

func solve_05_2(lines []line) int {
	grid := map[point]int{}
	for _, line := range lines {
		markLine(line, grid)
	}
	return countResult(grid)
}

func markLine(line line, grid map[point]int) {
	deltaX, deltaY := line.x2-line.x1, line.y2-line.y1
	dx, dy := direction(deltaX), direction(deltaY)
	n := Max(Abs(deltaX), Abs(deltaY))
	for x, y, i := line.x1, line.y1, 0; i <= n; x, y, i = x+dx, y+dy, i+1 {
		grid[point{x, y}]++
	}
}

func direction(delta int) int {
	switch {
	case delta < 0:
		return -1
	case delta > 0:
		return 1
	default:
		return 0
	}
}

func countResult(grid map[point]int) int {
	var result int
	for _, v := range grid {
		if v > 1 {
			result++
		}
	}
	return result
}

type line struct{ x1, y1, x2, y2 int }
type point struct{ x, y int }

func parse_05(s string) line {
	var x1, y1, x2, y2 int
	fmt.Sscanf(s, "%d,%d -> %d,%d", &x1, &y1, &x2, &y2)
	return line{x1, y1, x2, y2}
}

func main() {
	lines := GetParsedLines(parse_05)
	fmt.Println(solve_05_1(lines))
	fmt.Println(solve_05_2(lines))
}
